eventqueue: protect against enqueueing same Event twice

ianvernon opened the pull request:
Previously, if an Event is enqueued twice, the following panic would occur:

```
PANIC: eventqueue_test.go:187: EventQueueSuite.TestPanic

... Panic: close of closed channel (PC=0x42EFD4)

/usr/local/go/src/runtime/panic.go:522
  in gopanic
/usr/local/go/src/runtime/chan.go:342
  in closechan
eventqueue.go:188
  in EventQueue.Enqueue
eventqueue_test.go:197
  in EventQueueSuite.TestPanic
/usr/local/go/src/reflect/value.go:308
  in Value.Call
/usr/local/go/src/runtime/asm_amd64.s:1337
  in goexit
OOPS: 9 passed, 1 PANICKED
--- FAIL: Test (1.51s)
FAIL
coverage: 0.0% of statements in pkg/eventqueue
FAIL	github.com/cilium/cilium/pkg/eventqueue	1.550s
Makefile:129: recipe for target 'unit-tests' failed
make: *** [unit-tests] Error 1
```

To protect against this, do not allow enqueueing the same Event twice, by using
an atomic boolean which contains information about whether a given Event has
been enqueued or not.

Fixes: #8805

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8809)
<!-- Reviewable:end -->

Commit:
eventqueue: protect against enqueueing same Event twice

Previously, if an Event is enqueued twice, the following panic would occur:

```
PANIC: eventqueue_test.go:187: EventQueueSuite.TestPanic

... Panic: close of closed channel (PC=0x42EFD4)

/usr/local/go/src/runtime/panic.go:522
  in gopanic
/usr/local/go/src/runtime/chan.go:342
  in closechan
eventqueue.go:188
  in EventQueue.Enqueue
eventqueue_test.go:197
  in EventQueueSuite.TestPanic
/usr/local/go/src/reflect/value.go:308
  in Value.Call
/usr/local/go/src/runtime/asm_amd64.s:1337
  in goexit
OOPS: 9 passed, 1 PANICKED
--- FAIL: Test (1.51s)
FAIL
coverage: 0.0% of statements in pkg/eventqueue
FAIL	github.com/cilium/cilium/pkg/eventqueue	1.550s
Makefile:129: recipe for target 'unit-tests' failed
make: *** [unit-tests] Error 1
```

To protect against this, do not allow enqueueing the same Event twice, by using
an atomic boolean which contains information about whether a given Event has
been enqueued or not.

Fixes: #8805

Signed-off by: Ian Vernon <ian@cilium.io>

jrajahalme commented on pkg/eventqueue/eventqueue.go:
```diff
@@ -169,11 +176,16 @@ func (ev *Event) WasCancelled() bool {
+	if val := ev.enqueued.Load(); val == true {
+		return nil
+	}
+	ev.enqueued.Store(true)
```
This does not prevent two concurrent Enqueue()s for the same event, if they both Load `enqueued` before them both storing `true`. Maybe `if atomic.AddUint32(&ev.enqueued, 1) > 1 { return nil }` could be used instead?

ianvernon replied:
You are correct, done!

jrajahalme reviewed (changes requested):
Still a chance for a tight race.

aanm commented on pkg/eventqueue/eventqueue.go:
```diff
@@ -142,12 +147,13 @@ type eventStatistics struct {
 // NewEvent returns an Event with all fields initialized.
 func NewEvent(meta EventHandler) *Event {
-	return &Event{
+	ev := Event{
```
unnecessary change

aanm commented on pkg/eventqueue/eventqueue.go:
```diff
@@ -169,11 +175,15 @@ func (ev *Event) WasCancelled() bool {
 	}
 
+	// Events can only be enqueued once.
+	if atomic.AddInt32(&ev.enqueued, 1) > 1 {
```
ideally this should be a sync.Once, otherwise the event order is not being kept, i.e. if we enqueue the same event at the same time, the response of the 2nd queued might, and probably will, be returned before the 1st queued is executed.

ianvernon replied:
If we Enqueue the same event at the same time, one of the events will add to the int atomically, and the other one won't - how does that differ from a `sync.Once`? I can theoretically use a `sync.Once` since that is basically what I am using here (`sync.Once` is a wrapper around a mutex and `sync.Atomic` loading and storing a `uint32`.  

In other words, ordering will never be guaranteed in this case, it's a matter of who gets there first. I think that's fine. The `EventQueue` wasn't designed to have the same `Event` enqueued multiple times; safeguarding against that with some state which signifies whether it has been Enqueued is a needed fix, but I do not think that we need to ensure that if the same event is enqueued multiple times that we need to preserve the ordering of who calls Enqueue - that would mean that multiple goroutines would be Enqueueing the same event at the same time.

aanm replied:
>how does that differ from a sync.Once?

The sync.Once will wait until all other go routines being ran inside the Once are executed.

ianvernon replied:
I do not understand what advantages of waiting for the goroutines being ran inside the `sync.Once` are executed provides us is in this case. The critical section here is quite small as opposed to wrapping the contents of `Enqueue` itself within a `sync.Once`. Can you elaborate on what advantages that provides in this case?

aanm replied:
>Can you elaborate on what advantages that provides in this case?

Why is the function named "Enqueue" if we will return earlier in a 2nd, and following "enqueues" of an event that is still in the queue and still being processed? By earlier I meant that we will return even before the event was processed the first time.

ianvernon replied:
The first enqueue will process the event, the second enqueue will return a nil result immediately, which the user can check for. A `nil` return means that the Enqueue was a no-op. I can write that into the documentation. When would the ordering of returns matter in this case? I do not ever see a case in which a user would enqueue the same event twice across different goroutines. With this PR, we guard against that by returning a nil value for the second call to Enqueue. Can you provide an example where the ordering would matter? I feel like this is a semantic argument as opposed to one of correctness.

aanm replied:
It's weird to have a function called "enqueue" for which in reality is not enqueing events.

ianvernon replied:
The code was designed such that a given event should be enqueued only once. If an event is enqueued more than once, we return a signal to the user (`nil` result) that it was not enqueued. I could return an error as well from `Enqueue` indicating as such? If a user wants to `Enqueue` similar information again, they can call `NewEvent` and `Enqueue` that information. The path you are talking about is not a use-case that was envisioned when I wrote the code, nor is it how it is being used throughout Cilium.

jrajahalme replied:
To summarize: the `nil` return is an error condition the caller should check for.

aanm commented on pkg/eventqueue/eventqueue.go:
```diff
@@ -167,13 +172,19 @@ func (ev *Event) WasCancelled() bool {
 
+	// Events can only be enqueued once.
+	if atomic.AddInt32(&ev.enqueued, 1) > 1 {
+		return nil
```
This needs to be addressed somehow, it can be addressed in a different PR. If the caller of this function is **not** checking for the channel to be nil then doing `<-c` where `c` is a `nil` channel will deadlock forever.

jrajahalme replied:
Maybe document that the caller must check for `nil`, which is an error condition?

jrajahalme replied:
Note that `nil` return is done in other error conditions right above. Most likely all callers are already checking for `nil` return due to that.

aanm replied:
> Note that `nil` return is done in other error conditions right above. Most likely all callers are already checking for `nil` return due to that.

@jrajahalme yeah the problem is that they are not. I've checked.

ianvernon replied:
Fixed in all callpaths in latest commit. All call paths now check if the resultant channel is nil and return and / or log an appropriate error.

joestringer replied:
If this is an error case, then please return an error from the function. Golang has a ubiquitous standard for this, if there's an error, you return an error. Implementing functions that return error cases in a non-standard way is inviting future misuse.

ianvernon replied:
@joestringer fixed!

jrajahalme reviewed (approved):
Nit: document that the caller must check for error condition (`nil` return).

Commit:
eventqueue: return error if Enqueue fails

Given that an Enqueue may not occur due to the queue not being initialized
correctly / the Event being nil, or the Event has already been Enqueued before,
return an error which follows Golang conventions instead of encoding error state
in the returned value.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer reviewed (commented):
What's the context of this? Is there actually a codepath where this is occurring or was it observed while experimenting with new uses of the eventqueue?

IMO we should return this as an error and have the callers actually check/handle it as an error case. It seems weirdly out of the expected model of the eventqueue for the same event to be queued multiple times.
